### PR TITLE
irq: add irq interface header

### PR DESCRIPTION
**Pull request by cristiantpop**
This will abstract platform specific implementation

Signed-off-by: Cristian Pop <cristian.pop@analog.com>

### CODE REVIEW

File: include/irq.h [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+
+#ifndef IRQ_H_
+#define IRQ_H_
+#include <stdint.h>
+
+/******************************************************************************/
+/*************************** Types Declarations *******************************/
+/******************************************************************************/
+
+typedef struct irq_init_param {
```
Comment (adrimbarean) [inline include/irq.h]: Do we still want to keep typedefs for new implementations? The [Linux kernel coding style](https://www.kernel.org/doc/html/v4.10/process/coding-style.html#typedefs) we want to abide by, suggests against this type of usage.

**Comment by cristiantpop** [comment]
V2:
- remove typedefs

File: include/irq.h [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+
+/* Initialize the interrupt request peripheral. */
+int32_t irq_init(struct irq_desc **desc,
+		 const struct irq_init_param *param);
+
+/* Free the resources allocated by irq_init(). */
+int32_t irq_remove(struct irq_desc *desc);
+
+/* Registers a generic IRQ handling function */
+int32_t irq_register(struct irq_desc *desc, uint32_t irq_id,
```
Comment (dbogdan) [inline include/irq.h]: irq_desc already contains the irq_id information. Why do we need it here?

File: include/irq.h [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+	void		*extra;
+} irq_desc;
+
+/******************************************************************************/
+/************************ Functions Declarations ******************************/
+/******************************************************************************/
+
+/* Initialize the interrupt request peripheral. */
+int32_t irq_init(struct irq_desc **desc,
+		 const struct irq_init_param *param);
```
Comment (dbogdan) [inline include/irq.h]: What's the difference between irq_init() and irq_register()? Can't we have only one function? I fell that both are mandatory to be called.

File: projects/drivers/xilinx_platform/xilinx_platform_drivers.h [hunk 4]
```diff
@@ -65,6 +66,10 @@
 /*************************** Types Declarations *******************************/
 /******************************************************************************/
 
+struct xil_irq_desc {
+	XScuGic 		*gic;
+} xil_irq_desc;
```
Comment (dbogdan) [inline projects/drivers/xilinx_platform/xilinx_platform_drivers.h]: This won't compile on MicroBlaze platforms.

File: include/irq.h (hunk 3 above)
Comment (cristiantpop) [reply include/irq.h]: irq_init is like i2c_init, where the global interrupt system is initialized. 
irq_register is used to register a secondary interrupt like: UART, SPI, Timer...

File: include/irq.h (hunk 2 above)
Comment (cristiantpop) [reply include/irq.h]: "irq_desc already contains the irq_id": this is the global interrupt ID.
"irq_id" parameter in "irq_register" function refers to the id of the interrupt that needs to be register. For UART, an irq_id needs to be provided.

File: projects/drivers/xilinx_platform/xilinx_platform_drivers.h (hunk 4 above)
Comment (dbogdan) [reply projects/drivers/xilinx_platform/xilinx_platform_drivers.h]: We should at least add "#ifdef _XPARAMETERS_PS_H_".

File: include/irq.h (hunk 2 above)
Comment (dbogdan) [reply include/irq.h]: The naming is a little bit confusing for me. :)

**Comment by cristiantpop** [comment]
V3:
- Add "#ifdef XPARAMETERS_PS_H".

**Comment by mirceacaprioru** [comment]
if there are no other remarks will merge later today.

**Pull request closed without merging**